### PR TITLE
Created an inital guide on how to make audiostreams

### PR DESCRIPTION
**Pull request by hungrymonkey**
I probably will have to make changes. Please comment on this page.

### CODE REVIEW

**Comment by hungrymonkey** [comment]
there might be bugs in the sample code.

here is the link to the code i copied over to the docs

https://github.com/hungrymonkey/godot/tree/audiostream_doc/modules/tone_audiostream

**Comment by hungrymonkey** [comment]
@mhilbrunner I am debating a few things

First, the example code does not compile anymore because upstream change https://github.com/godotengine/godot/commit/8a9f1c2a5d7364016b9c67dc158557607f6de4bd

I have to add get_length function to audiostream. I would probably just return zero.

Second, I realize I missed an important as mention by reduz in this pull request fact that AudioStreamPlayback is managed by the audio thread. Therefore, heavy io is forbidden.

https://github.com/godotengine/godot/pull/15967#issuecomment-359808188

File: development/cpp/custom_audiostreams.rst [hunk 1]
```diff
@@ -0,0 +1,345 @@
+.. _custom_audiostreams:
+
+Custom AudioStreams
+===================
+
+Introduction
+------------
+
+AudioStream is the base class of all audio emitting objects. 
+AudioStreamPlayer binds onto an AudioStream to emit pcm data
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: PCM (captalized)

File: development/cpp/custom_audiostreams.rst [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+===================
+
+Introduction
+------------
+
+AudioStream is the base class of all audio emitting objects. 
+AudioStreamPlayer binds onto an AudioStream to emit pcm data 
+into an AudioServer which manages audio drivers.
+
+All audio resources requires two audio based classes; AudioStream
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: require (not **requires**)

File: development/cpp/custom_audiostreams.rst [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+------------
+
+AudioStream is the base class of all audio emitting objects. 
+AudioStreamPlayer binds onto an AudioStream to emit pcm data 
+into an AudioServer which manages audio drivers.
+
+All audio resources requires two audio based classes; AudioStream 
+and AudioStreamPlayback. As a data container, AudioStream contains 
+the resource and exposes itself to GDScript. AudioStream references 
+its own internal custom AudioStreamPlayback which translate
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: translate**s**

File: development/cpp/custom_audiostreams.rst [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+AudioStreamPlayer binds onto an AudioStream to emit pcm data 
+into an AudioServer which manages audio drivers.
+
+All audio resources requires two audio based classes; AudioStream 
+and AudioStreamPlayback. As a data container, AudioStream contains 
+the resource and exposes itself to GDScript. AudioStream references 
+its own internal custom AudioStreamPlayback which translate 
+AudioStream into PCM data.
+
+This doc assumes the reader knows how to create C++ Modules. If not, refer to this guide
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: **m**odules, also I'd change it to "this guide: "

File: development/cpp/custom_audiostreams.rst [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+- Binding external libraries (like Wwise, FMOD, etc).
+- Adding custom audio queues
+- Adding support for more audio formats
+
+Create an AudioStream
+---------------------
+
+
+An AudioStream consists of three components; data container, stream name,
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: ":" instead of ";" after components

File: development/cpp/custom_audiostreams.rst [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+- Adding custom audio queues
+- Adding support for more audio formats
+
+Create an AudioStream
+---------------------
+
+
+An AudioStream consists of three components; data container, stream name, 
+and an AudioStreamPlayback friend class generator. Audio data can be 
+loaded a number of ways such as an internal counter for a tone generator,
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: loaded **in** a number of ways such as **with** an internal counter ...

File: development/cpp/custom_audiostreams.rst [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+---------------------
+
+
+An AudioStream consists of three components; data container, stream name, 
+and an AudioStreamPlayback friend class generator. Audio data can be 
+loaded a number of ways such as an internal counter for a tone generator, 
+internal/external buffer, or a file reference.
+
+
+Some AudioStreams need to be stateless such as objects loaded from
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: such as = like?

File: development/cpp/custom_audiostreams.rst [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+References:
+~~~~~~~~~~~
+
+-  `servers/audio/audio_stream.h <https://github.com/godotengine/godot/blob/master/servers/audio/audio_stream.h>`__
+
+
+Create an AudioStreamPlayback
+-----------------------------
+
+AudioStreamPlayer uses ``mix`` callback to obtain pcm data. The callback must match sample rate and fill the buffer.
```
Comment (mhilbrunner) [inline development/cpp/custom_audiostreams.rst]: Capitalize PCM

**Comment by mhilbrunner** [comment]
@hungrymonkey Commented on a few things, reviewed the whole thing. Looks good so far! Thanks for working on this.

If you are interested in fixing the issues you pointed out yourself, I think this would be a great addition for the docs.

File: development/cpp/custom_audiostreams.rst (hunk 7 above)
Comment (hungrymonkey) [reply development/cpp/custom_audiostreams.rst]: basically yea.

**Comment by hungrymonkey** [comment]
I am going add all of those changes and add whatever suggestions i made above

**Comment by hungrymonkey** [comment]
@mhilbrunner check if i missed anything

**Comment by hungrymonkey** [comment]
wait i made a small mistake with the code

**Comment by hungrymonkey** [comment]
done

**Comment by hungrymonkey** [comment]
done

**Comment by mhilbrunner** [comment]
Looks good. Lets give the rest of the docs team a little more time to spot things I missed, but it looks good to go to me. Thanks again for the contribution!

**Comment by NathanLovato** [comment]
All good. Let's merge so that other people can proof it and submit PRs if need be :)

**Comment by hungrymonkey** [comment]
hmmm, it takes around 30 minute from merge to show up

http://docs.godotengine.org/en/latest/development/cpp/custom_audiostreams.html

**Comment by NathanLovato** [comment]
That's normal don't worry - the engine behind the docs is pretty slow, if you've built your page locally you should've seen that too

**Pull request merged**